### PR TITLE
Add extensible smgr slot for other storage format.

### PR DESCRIPTION
**Pull request by wenchaozhang-123**
closes: #ISSUE_Number

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to @cloudberrydb/dev team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/include/storage/smgr.h [hunk 1]
```diff
@@ -183,4 +181,7 @@ extern PGDLLIMPORT file_truncate_hook_type file_truncate_hook;
 typedef void (*file_unlink_hook_type)(RelFileNodeBackend rnode);
 extern PGDLLIMPORT file_unlink_hook_type file_unlink_hook;
 
+extern f_smgr smgrsw[];
+extern int NSmgr;
+extern int CurSmgr;
```
Comment (my-ship-it) [inline src/include/storage/smgr.h:186]: Could we add a function like add_reloption_kind to allocate Smgr?

File: src/include/storage/smgr.h [hunk 2]
```diff
@@ -23,13 +23,11 @@
 #include "storage/dbdirnode.h"
 #include "utils/relcache.h"
 
-typedef enum SMgrImplementation
-{
-	SMGR_MD = 0,
-	SMGR_AO = 1
-} SMgrImpl;
```
Comment (my-ship-it) [inline src/include/storage/smgr.h:30]: SMGR_MAX = lengthof(smgrsw)

File: src/backend/catalog/storage.c [hunk 3]
```diff
@@ -157,7 +157,7 @@ RelationCreateStorage(RelFileNode rnode, char relpersistence, SMgrImpl smgr_whic
 	pending->relnode.isTempRelation = backend == TempRelBackendId;
 	pending->atCommit = false;	/* delete if abort */
 	pending->nestLevel = GetCurrentTransactionNestLevel();
-	pending->relnode.smgr_which = smgr_which;
+	pending->relnode.smgr_which = srel->smgr_which;;
```
Comment (my-ship-it) [inline src/backend/catalog/storage.c:160]: ;; ?

Comment (wenchaozhang-123) [reply src/backend/catalog/storage.c]: solved.

**Review by my-ship-it — DISMISSED** [review]

File: src/backend/storage/smgr/smgr.c [hunk 4]
```diff
@@ -754,3 +780,16 @@ AtEOXact_SMgr(void)
 		smgrclose(rel);
 	}
 }
+
+SMgrImpl
+add_smgr_kind(void)
```
Comment (gongxun0928) [inline src/backend/storage/smgr/smgr.c:785]: it better to add some comments for it. 
To use custom smgr,  it should first call add_smgr_kind() to get a slot, and then initialize the slot.

Comment (wenchaozhang-123) [reply src/backend/storage/smgr/smgr.c:785]: OK

File: src/backend/storage/smgr/smgr.c [hunk 5]
```diff
@@ -754,3 +780,20 @@ AtEOXact_SMgr(void)
 		smgrclose(rel);
 	}
 }
+
+/*
+ * When need to add a new storage format smgr in extension, we should
+ * call add_smgr_kind to get a slot, then init init the slot.
```
Comment (gongxun0928) [inline src/backend/storage/smgr/smgr.c:786]: typo: init init

File: src/backend/storage/smgr/smgr.c [hunk 6]
```diff
@@ -754,3 +780,20 @@ AtEOXact_SMgr(void)
...
+{
+	/* Now, only support NSmgr smgrs */
+	if (last_assigned_smgr_kind >= NSmgr)
+		ereport(ERROR,
+				(errcode(ERRCODE_PROGRAM_LIMIT_EXCEEDED),
+				 errmsg("extensible smgr types limit exceeded")));
+
+	last_assigned_smgr_kind++;
+	return (SMgrImpl) last_assigned_smgr_kind;
+}
```
Comment (gfphoenix78) [inline src/backend/storage/smgr/smgr.c:799]: Is the smgr id local or global?

Comment (wenchaozhang-123) [reply src/backend/storage/smgr/smgr.c:799]: No, it's session level

File: src/backend/catalog/storage.c [hunk 7]
```diff
@@ -198,6 +198,12 @@ void
 RelationDropStorage(Relation rel)
 {
 	PendingRelDelete *pending;
+	SMgrRelation srel;
+
+	srel = smgropen(rel->rd_node, rel->rd_backend == TempRelBackendId ?
```
Comment (HuSen8891) [inline src/backend/catalog/storage.c:203]: use rel->rd_backend directly? smgropen(rel->rd_node, rel->rd_backend, .....)

Comment (wenchaozhang-123) [reply src/backend/catalog/storage.c:203]: OK，that's right

**Pull request closed without merging**